chapter extraction as segmentation

darolfes opened the pull request:
This pr extends the VideoSegmenter with the ability to extract available chapter information from video tracks and use this information for the track segmentation.

darolfes committed:
chapter extraction as segmentation

Arnei commented on modules/videosegmenter-ffmpeg/src/main/java/org/opencastproject/videosegmenter/ffmpeg/VideoSegmenterServiceImpl.java:
```diff
@@ -318,6 +346,39 @@ public void updated(Dictionary<String, ?> properties) throws ConfigurationExcept
+        useChapterIfAvailable = Boolean.parseBoolean(value);
+        logger.info("Use Chapters if available is set to {}", useChapterIfAvailable);
+      } catch (Exception e) {
+        logger.warn("Found illegal value '{}' for videosegmenter's option: {}",
```
I think it would be better to log an error here and throw a ConfigurationException.
I am aware the current code is in line with the rest of the `updated()` method, maybe there is a reason for why throwing a ConfigurationException is problematic somehow. But if it isn't, i'd throw it. Really let the user know about their typo.

Arnei commented on modules/videosegmenter-ffmpeg/src/main/java/org/opencastproject/videosegmenter/ffmpeg/VideoSegmenterServiceImpl.java:
```diff
@@ -318,6 +346,39 @@ public void updated(Dictionary<String, ?> properties) throws ConfigurationExcept
+
+        useChapterMimeTypes = mts;
+      } catch (Exception e) {
+        logger.warn("Found illegal value '{}' for videosegmenter's option: {}",
```
See comment about ConfigurationException above.

Arnei commented on modules/videosegmenter-ffmpeg/src/main/java/org/opencastproject/videosegmenter/ffmpeg/VideoSegmenterServiceImpl.java:
```diff
@@ -379,252 +438,498 @@ protected Catalog segment(Job job, Track track)
-      // optimization loop to get a segmentation with a number of segments close
-      // to the desired number of segments
-      while (!endOptimization) {
+    logger.info("Detecting chapters using command: {}", (Object) command);
```
```suggestion
    logger.debug("Detecting chapters using command: {}", (Object) command);
```

Arnei commented on modules/videosegmenter-ffmpeg/src/main/java/org/opencastproject/videosegmenter/ffmpeg/VideoSegmenterServiceImpl.java:
```diff
@@ -379,252 +438,498 @@ protected Catalog segment(Job job, Track track)
+      // Being processing of title
+      else if (state == 5) {
+        if (!line.startsWith("title=")) {
+          throw new ParseException("Failed to parse FFMETADATA: CHAPTER title field missing", lineNumber);
```
Is the title field required? I can't find anything about it in the documentation.

Arnei replied:
Couldn't find anything on whether the title field is required or not. I am willing to give David the benefit of the doubt here and just accept this.

Arnei commented on modules/videosegmenter-ffmpeg/src/main/java/org/opencastproject/videosegmenter/ffmpeg/VideoSegmenterServiceImpl.java:
```diff
@@ -45,11 +45,15 @@
 import org.opencastproject.videosegmenter.api.VideoSegmenterService;
 import org.opencastproject.workspace.api.Workspace;
 
+import com.entwinemedia.fn.data.Opt;
```
We would like to eventually get rid of the entwine classes. Please use `java.util.Optional` or something alike.

Arnei reviewed (changes requested):
I imagine a unit test for this would be really nice.

I wonder how this is expected to be used? Are there capture agents that include chapter information in their video files? Depending on who this is relevant to, the documentation should be amended.

Code works from what I tested, the comments below are minor stuff.

gregorydlogan commented:
Any progress on this?

Arnei committed:
Remove entwine from VideoSegmenterServiceImpl.java

Replaced entwine Opt with java Optional

Arnei committed:
Throw exceptions on misconfigurations in VideoSegmenterServiceImpl.java

Instead of logging a warning message, we now throw a ConfigurationException instead in order to properly inform the user.

Arnei reviewed (approved):
Fixed my review myself. Should be good to go now.

darolfes commented:
@Arnei Sorry, I had little time to come back to this.
I still have a patch that makes the title field optional (but parses it if it is present). I could push it if you want. However, it's still completely untested.

Arnei commented:
> @Arnei Sorry, I had little time to come back to this. I still have a patch that makes the title field optional (but parses it if it is present). I could push it if you want. However, it's still completely untested.

Oh hey, nice hearing from you. Didn't expect you to come back to this :D
Good to know that the title field should be optional after all. If you have code to that effect, please do contribute it. Shouldn't be that much work to test it.

darolfes committed:
Make title field for chapter extraction of VideoSegmenterServiceImpl.java optional

darolfes commented:
Pushed it. Hopefully, it works. Looking it over, I am unsure of the `continue`s of the changed states. The force push adds one to state 7.
Edit: Wrong line, now it does.

Arnei commented:
Sooo, I tried to create a FFMETADATA file with a chapter without a title, looking like this:
```
;FFMETADATA1
major_brand=isom
minor_version=512
compatible_brands=isomiso2avc1mp41
encoder=Lavf58.76.100

[CHAPTER]
TIMEBASE=1/1000
START=1
END=6000
title=The Pledge

[CHAPTER]
TIMEBASE=1/1000
START=6001
END=12000

[CHAPTER]
TIMEBASE=1/1000
START=12001
END=20000
title= The Prestige
```

Trying to add this file to a video results in wrong chapter information:

```
;FFMETADATA1
major_brand=isom
minor_version=512
compatible_brands=isomiso2avc1mp41
encoder=Lavf58.76.100
[CHAPTER]
TIMEBASE=1/1000
START=0
END=12001
title=The Pledge
[CHAPTER]
TIMEBASE=1/1000
START=12001
END=20000
title= The Prestige
[CHAPTER]
TIMEBASE=1/10000000
START=120010000
END=647000000
title= The Prestige
```

So maybe the title field is required after all? :')

darolfes commented:
Interesting, I looked at my test files, and one file produces the following ffmetadata:

```
;FFMETADATA1
encoder=Lavf60.3.100
[CHAPTER]
TIMEBASE=1/1000000000
START=1000000
END=10000000000
[CHAPTER]
TIMEBASE=1/1000000000
START=10001000000
END=40000000000
title=Middle\
Test
[CHAPTER]
TIMEBASE=1/1000000000
START=40001000000
END=60000000000
title=End
```

The title field is missing. Maybe it depends on the chosen container format (this one is webm), or it only "works" for the first chapter, which would indicate a bug in FFmpeg (either by accepting/processing corrupt ffmetadata files or by incorrectly setting the chapter information in the video files when missing titles are present). VLC displays the first chapter as `Chapter 0`.

darolfes commented:
After reproducing the video file:
 Are there already chapters defined in your input file?
 This led to problems for me. A mapping with map_chapters is probably needed in this case (did not test this). With input files without chapters, I can reproduce the video file without a problem.

Arnei commented:
> Maybe it depends on the chosen container format (this one is webm)

Interesting. My original test was with a `mp4` container. Another test I did with `mp4` also failed. But with a `webm` it works (at least the title information is preserved correctly, webm seems to default to changing the timebase back to nanoseconds for whatever reason)

> Are there already chapters defined in your input file?

All files I tried to add chapter information to did not previously have chapter information.

Arnei commented on modules/videosegmenter-ffmpeg/src/main/java/org/opencastproject/videosegmenter/ffmpeg/VideoSegmenterServiceImpl.java:
```diff
@@ -379,252 +453,506 @@ protected Catalog segment(Job job, Track track)
+
+    String line = reader.readLine();
+    int lineNumber = 1;
+    while (null != line) {
```
Since you mentioned being unsure about `continue`s, I checked the loop again. This can cause the last chapter in the file to not be added, if the last line in the FFMETADATA file belongs to the last chapter. The loop will parse the last line, read the next line which is null (as it does not exist), and exit the loop before it can get to the `state 7` part. Do you want to fix this or should I?

darolfes replied:
A similar problem exists with state 5 as it can now directly lead into the new state 7. Additionally a ParseException would probably be better instead of a NullPointerException for unexpected file ends.

darolfes replied:
Will push an update.

darolfes replied:
Hopefully this fixes the problem: [55163d..bd6b50](https://github.com/opencast/opencast/compare/55163d681e85628a8645d636adfd00dd3cd44321..bd6b50c24c2372e8c20f6d2234dd03fb32fdee84)

Arnei replied:
This opens up the possibility of `line` being `null` during the loop, which can now cause NullPointerExceptions with e.g. `startsWith("")`

darolfes replied:
This only affects the comments/empty line if-branch or did I overlook something else (as all states except 7 are never visited if line == null)?
[Fix](https://github.com/opencast/opencast/compare/bd6b50c24c2372e8c20f6d2234dd03fb32fdee84..6cd02b092cc1286f3b9d13ee9b2293f3f8f7a68b)

Arnei replied:
Seems so, I just saw that it would be a problem in at least one case ^^.
